internal/keepers/filter.go

```diff
@@ -0,0 +1,176 @@
+package keepers
+
+import (
+	"context"
+	"log"
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/smartcontractkit/ocr2keepers/pkg/types"
+)
+
+type reportCoordinator struct {
+	logger       *log.Logger
+	registry     types.Registry
+	logs         types.PerformLogProvider
+	idBlocks     *cache[bool] // should clear out when the next perform with this id occurs
+	activeKeys   *cache[bool] // should clear when next perform with blocknumber and id occurs
+	cacheCleaner *intervalCacheCleaner[bool]
+	starter      sync.Once
+	chStop       chan struct{}
+}
+
+func newReportCoordinator(r types.Registry, s time.Duration, cacheClean time.Duration, logs types.PerformLogProvider, logger *log.Logger) *reportCoordinator {
+	c := &reportCoordinator{
+		logger:     logger,
+		registry:   r,
+		logs:       logs,
+		idBlocks:   newCache[bool](s),
+		activeKeys: newCache[bool](time.Hour), // 1 hour allows the cleanup routine to clear stale data
+		chStop:     make(chan struct{}),
+	}
+
+	cl := &intervalCacheCleaner[bool]{
+		Interval: cacheClean,
+		stop:     make(chan struct{}),
+	}
+
+	c.cacheCleaner = cl
+
+	runtime.SetFinalizer(c, func(srv *reportCoordinator) { srv.stop() })
+
+	c.start()
+
+	return c
+}
+
+// Add adds the provided upkeep to the filter.
+func (rc *reportCoordinator) Add(key types.UpkeepKey) error {
+	id, err := rc.registry.IdentifierFromKey(key)
+	if err != nil {
+		return err
+	}
+
+	rc.idBlocks.Set(string(id), true, defaultExpiration)
+
+	return nil
+}
+
+// Filter returns a filter function that removes upkeep keys that apply to this
+// filter
+func (rc *reportCoordinator) Filter() func(types.UpkeepKey) bool {
+	return func(key types.UpkeepKey) bool {
+		id, err := rc.registry.IdentifierFromKey(key)
+		if err != nil {
+			return false
+		}
+
+		if _, ok := rc.idBlocks.Get(string(id)); ok {
+			return false
+		}
+
+		return true
+	}
+}
+
+func (rc *reportCoordinator) Accept(key types.UpkeepKey) {
+	rc.activeKeys.Set(string(key), false, defaultExpiration)
+}
+
+func (rc *reportCoordinator) IsTransmitting(key types.UpkeepKey) bool {
+	// key is assumed to be transmitted if it doesn't exist in cache or if it
+	// is no longer waiting for a transaction to be transmitted
+	inLogs, ok := rc.activeKeys.Get(string(key))
+	return ok && inLogs
+}
+
+func (rc *reportCoordinator) start() {
+	rc.starter.Do(func() {
+		go rc.run()
+		go rc.cacheCleaner.Run(rc.idBlocks)
+		go rc.cacheCleaner.Run(rc.activeKeys)
+	})
+}
+
+func (rc *reportCoordinator) stop() {
+	rc.chStop <- struct{}{}
+	rc.cacheCleaner.stop <- struct{}{}
+	rc.cacheCleaner.stop <- struct{}{}
+}
+
+func (rc *reportCoordinator) run() {
+	// TODO: handle panics by restarting this process
+
+	cadence := time.Second
+	timer := time.NewTimer(cadence)
+
+	for {
+
+		select {
+		case <-timer.C:
+			startTime := time.Now()
+			keys := rc.activeKeys.Keys()
+			logs, _ := rc.logs.PerformLogs(context.Background())
```

Comment:
note: this needs to look back some range of blocks (i think depends on performLockoutWindow and deltaStage * n). Would be good to have that logic here

Reply:
Let's also bring that up in discussion. I'm guessing the lookback is chain specific but can it be an implementation detail of the PerformLogsProvider?